Change old locations services in favor of new LocationEngine abstraction

Some activities which are using old locations services (`GeocodingWidgetActivity` and `GeocodingServiceActivity`) are crashing when calling `FusedLocationProviderService.requestLocationUpdates` or `FusedLocationProviderService.getLastLocation` methods.

> FATAL EXCEPTION: main
                                                                                     Process: com.mapbox.services.android.testapp, PID: 26935
                                                                                     java.lang.RuntimeException: Unable to start activity ComponentInfo{com.mapbox.services.android.testapp/com.mapbox.services.android.testapp.geocoding.GeocodingServiceActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.location.Location com.mapzen.android.lost.internal.FusedLocationProviderService.getLastLocation(com.mapzen.android.lost.api.LostApiClient)' on a null object reference
                                                                                         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
                                                                                         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                                                                                         at android.app.ActivityThread.-wrap11(ActivityThread.java)
                                                                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                                                                                         at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                         at android.os.Looper.loop(Looper.java:148)
                                                                                         at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                                      Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.location.Location com.mapzen.android.lost.internal.FusedLocationProviderService.getLastLocation(com.mapzen.android.lost.api.LostApiClient)' on a null object reference
                                                                                         at com.mapzen.android.lost.internal.FusedLocationProviderApiImpl.getLastLocation(FusedLocationProviderApiImpl.java:100)
                                                                                         at com.mapbox.services.android.testapp.geocoding.GeocodingServiceActivity.getLastLocation(GeocodingServiceActivity.java:155)
                                                                                         at com.mapbox.services.android.testapp.geocoding.GeocodingServiceActivity.onStart(GeocodingServiceActivity.java:148)
                                                                                         at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1237)
                                                                                         at android.app.Activity.performStart(Activity.java:6253)
                                                                                         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2379)
                                                                                         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
                                                                                         at android.app.ActivityThread.-wrap11(ActivityThread.java) 
                                                                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
                                                                                         at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                         at android.os.Looper.loop(Looper.java:148) 
                                                                                         at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                                         at java.lang.reflect.Method.invoke(Native Method) 
                                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 

Use new `LocationEngine` abstraction to fix those crashes.

--- a/mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java
+++ b/mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java
@@ -186,6 +186,7 @@ public class GeocodingServiceActivity extends AppCompatActivity implements Locat
 
   @Override
   public void onLocationChanged(Location location) {
+    // Unused on purpose
   }
 
   @Override

--- a/mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java
+++ b/mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java
@@ -152,11 +152,11 @@ public class GeocodingServiceActivity extends AppCompatActivity implements Locat
   protected void onResume() {
     super.onResume();
     if (locationEngine != null && locationEngine.isConnected()) {
-      getLastLocation();
+      obtainLastAddress();
     }
   }
 
-  private void getLastLocation() {
+  private void obtainLastAddress() {
     // Gets the best and most recent location currently available, which may be null
     // in rare cases when a location is not available.
 
@@ -181,14 +181,11 @@ public class GeocodingServiceActivity extends AppCompatActivity implements Locat
 
   @Override
   public void onConnected() {
-    getLastLocation();
+    obtainLastAddress();
   }
 
   @Override
   public void onLocationChanged(Location location) {
-    if (location != null) {
-      getLastLocation();
-    }
   }
 
   @Override

--- a/mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java
+++ b/mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java
@@ -1,5 +1,6 @@
 package com.mapbox.services.android.testapp.geocoding;
 
+import android.content.Context;
 import android.content.Intent;
 import android.location.Location;
 import android.os.Bundle;
@@ -13,49 +14,52 @@ import android.widget.TextView;
 import android.widget.Toast;
 
 import com.mapbox.services.android.geocoder.AndroidGeocoder;
+import com.mapbox.services.android.telemetry.location.AndroidLocationEngine;
+import com.mapbox.services.android.telemetry.location.LocationEngine;
+import com.mapbox.services.android.telemetry.location.LocationEngineListener;
 import com.mapbox.services.android.testapp.R;
 import com.mapbox.services.android.testapp.geocoding.service.Constants;
 import com.mapbox.services.android.testapp.geocoding.service.FetchAddressIntentService;
-import com.mapzen.android.lost.api.LocationServices;
-import com.mapzen.android.lost.api.LostApiClient;
+
+import java.lang.ref.WeakReference;
 
 /**
  * This activity is inspired by the stock Android Geocoder sample code in
  * https://github.com/googlesamples/android-play-location/tree/master/LocationAddress
  * <p>
- * In this sample, we show how to use Mapbox' Geocoder simply replacing the
+ * In this sample, we show how to use Mapbox's Geocoder simply replacing the
  * android.location.Geocoder object with com.mapbox.services.android.geocoder.AndroidGeocoder.
- * To simplify the code, we've replaced Google Play Services with LOST.
+ * To simplify the code, we've replaced Google Play Services with Android.
  */
-public class GeocodingServiceActivity extends AppCompatActivity {
+public class GeocodingServiceActivity extends AppCompatActivity implements LocationEngineListener {
 
-  protected static final String ADDRESS_REQUESTED_KEY = "address-request-pending";
-  protected static final String LOCATION_ADDRESS_KEY = "location-address";
+  private static final String ADDRESS_REQUESTED_KEY = "address-request-pending";
+  private static final String LOCATION_ADDRESS_KEY = "location-address";
 
   /**
-   * Provides the entry point to LOST services.
+   * Provides the entry point to location services.
    */
-  protected LostApiClient lostApiClient;
+  private LocationEngine locationEngine;
 
   /**
    * Represents a geographical location.
    */
-  protected Location lastLocation;
+  private Location lastLocation;
 
   /**
    * Tracks whether the user has requested an address. Becomes true when the user requests an
    * address and false when the address (or an error message) is delivered.
    * The user requests an address by pressing the Fetch Address button. This may happen
-   * before LostApiClient connects. This activity uses this boolean to keep track of the
+   * before location engine connects. This activity uses this boolean to keep track of the
    * user's intent. If the value is true, the activity tries to fetch the address as soon as
-   * LostApiClient connects.
+   * location engine connects.
    */
-  protected boolean addressRequested;
+  private static boolean addressRequested;
 
   /**
    * The formatted location address.
    */
-  protected String addressOutput;
+  private static String addressOutput;
 
   /**
    * Receiver registered with this activity to get the response from FetchAddressIntentService.
@@ -65,36 +69,37 @@ public class GeocodingServiceActivity extends AppCompatActivity {
   /**
    * Displays the location address.
    */
-  protected TextView locationAddressTextView;
+  private TextView locationAddressTextView;
 
   /**
    * Visible while the address is being fetched.
    */
-  ProgressBar progressBar;
+  private ProgressBar progressBar;
 
   /**
    * Kicks off the request to fetch an address when pressed.
    */
-  Button fetchAddressButton;
+  private Button fetchAddressButton;
 
   @Override
   public void onCreate(Bundle savedInstanceState) {
     super.onCreate(savedInstanceState);
     setContentView(R.layout.activity_geocoding_service);
 
-    resultReceiver = new AddressResultReceiver(new Handler());
-
     locationAddressTextView = (TextView) findViewById(R.id.location_address_view);
     progressBar = (ProgressBar) findViewById(R.id.progress_bar);
     fetchAddressButton = (Button) findViewById(R.id.fetch_address_button);
+
+    resultReceiver = new AddressResultReceiver(new Handler(), this, locationAddressTextView, progressBar,
+      fetchAddressButton);
 
     // Set defaults, then update using values stored in the Bundle.
     addressRequested = false;
     addressOutput = "";
     updateValuesFromBundle(savedInstanceState);
 
-    updateUiWidgets();
-    buildLostApiClient();
+    updateUiWidgets(addressRequested, progressBar, fetchAddressButton);
+    buildAndroidLocationEngine();
   }
 
   /**
@@ -111,48 +116,51 @@ public class GeocodingServiceActivity extends AppCompatActivity {
       // and stored in the Bundle. If it was found, display the address string in the UI.
       if (savedInstanceState.keySet().contains(LOCATION_ADDRESS_KEY)) {
         addressOutput = savedInstanceState.getString(LOCATION_ADDRESS_KEY);
-        displayAddressOutput();
+        displayAddressOutput(locationAddressTextView);
       }
     }
   }
 
   /**
-   * Builds a LostApiClient.
+   * Builds Android location engine
    */
-  protected synchronized void buildLostApiClient() {
-    lostApiClient = new LostApiClient.Builder(this).build();
+  private synchronized void buildAndroidLocationEngine() {
+    locationEngine = AndroidLocationEngine.getLocationEngine(this);
+    locationEngine.addLocationEngineListener(this);
+    locationEngine.activate();
   }
 
   /**
    * Runs when user clicks the Fetch Address button. Starts the service to fetch the address if
-   * LostApiClient is connected.
+   * location engine is connected.
    */
   public void fetchAddressButtonHandler(View view) {
-    // We only start the service to fetch the address if LostApiClient is connected.
-    if (lostApiClient.isConnected() && lastLocation != null) {
+    // We only start the service to fetch the address if location engine is connected.
+    if (locationEngine.isConnected() && lastLocation != null) {
       startIntentService();
     }
 
-    // If LostApiClient isn't connected, we process the user's request by setting
-    // addressRequested to true. Later, when LostApiClient connects, we launch the service to
+    // If location engine isn't connected, we process the user's request by setting
+    // addressRequested to true. Later, when location engine connects, we launch the service to
     // fetch the address. As far as the user is concerned, pressing the Fetch Address button
     // immediately kicks off the process of getting the address.
     addressRequested = true;
-    updateUiWidgets();
+    updateUiWidgets(addressRequested, progressBar, fetchAddressButton);
   }
 
   @Override
-  protected void onStart() {
-    super.onStart();
-    lostApiClient.connect();
-    getLastLocation();
+  protected void onResume() {
+    super.onResume();
+    if (locationEngine != null && locationEngine.isConnected()) {
+      getLastLocation();
+    }
   }
 
   private void getLastLocation() {
     // Gets the best and most recent location currently available, which may be null
     // in rare cases when a location is not available.
-    //noinspection MissingPermission
-    lastLocation = LocationServices.FusedLocationApi.getLastLocation(lostApiClient);
+
+    lastLocation = locationEngine.getLastLocation();
     if (lastLocation != null) {
       // Determine whether a Geocoder is available.
       if (!AndroidGeocoder.isPresent()) {
@@ -161,10 +169,10 @@ public class GeocodingServiceActivity extends AppCompatActivity {
       }
 
       // It is possible that the user presses the button to get the address before the
-      // LostApiClient object successfully connects. In such a case, addressRequested
+      // location engine object successfully connects. In such a case, addressRequested
       // is set to true, but no attempt is made to fetch the address (see
       // fetchAddressButtonHandler()) . Instead, we start the intent service here if the
-      // user has requested an address, since we now have a connection to LostApiClient.
+      // user has requested an address, since we now have a connection to location engine.
       if (addressRequested) {
         startIntentService();
       }
@@ -172,10 +180,23 @@ public class GeocodingServiceActivity extends AppCompatActivity {
   }
 
   @Override
-  protected void onStop() {
-    super.onStop();
-    if (lostApiClient.isConnected()) {
-      lostApiClient.disconnect();
+  public void onConnected() {
+    getLastLocation();
+  }
+
+  @Override
+  public void onLocationChanged(Location location) {
+    if (location != null) {
+      getLastLocation();
+    }
+  }
+
+  @Override
+  protected void onPause() {
+    super.onPause();
+    if (locationEngine != null) {
+      locationEngine.removeLocationEngineListener(this);
+      locationEngine.deactivate();
     }
   }
 
@@ -183,7 +204,7 @@ public class GeocodingServiceActivity extends AppCompatActivity {
    * Creates an intent, adds location data to it as an extra, and starts the intent service for
    * fetching an address.
    */
-  protected void startIntentService() {
+  private void startIntentService() {
     // Create an intent for passing to the intent service responsible for fetching the address.
     Intent intent = new Intent(this, FetchAddressIntentService.class);
 
@@ -202,28 +223,28 @@ public class GeocodingServiceActivity extends AppCompatActivity {
   /**
    * Updates the address in the UI.
    */
-  protected void displayAddressOutput() {
-    locationAddressTextView.setText(addressOutput);
+  private static void displayAddressOutput(TextView view) {
+    view.setText(addressOutput);
   }
 
   /**
    * Toggles the visibility of the progress bar. Enables or disables the Fetch Address button.
    */
-  private void updateUiWidgets() {
+  private static void updateUiWidgets(boolean addressRequested, ProgressBar progressBar, Button button) {
     if (addressRequested) {
       progressBar.setVisibility(ProgressBar.VISIBLE);
-      fetchAddressButton.setEnabled(false);
+      button.setEnabled(false);
     } else {
       progressBar.setVisibility(ProgressBar.GONE);
-      fetchAddressButton.setEnabled(true);
+      button.setEnabled(true);
     }
   }
 
   /**
-   * Shows a toast with the given text.
+   * Shows a toast with the given message.
    */
-  protected void showToast(String text) {
-    Toast.makeText(this, text, Toast.LENGTH_SHORT).show();
+  private static void showToast(Context context, String message) {
+    Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
   }
 
   @Override
@@ -239,9 +260,19 @@ public class GeocodingServiceActivity extends AppCompatActivity {
   /**
    * Receiver for data sent from FetchAddressIntentService.
    */
-  class AddressResultReceiver extends ResultReceiver {
-    public AddressResultReceiver(Handler handler) {
+  private static class AddressResultReceiver extends ResultReceiver {
+    private final WeakReference<Context> contextReference;
+    private final WeakReference<TextView> textViewReference;
+    private final WeakReference<ProgressBar> progressBarReference;
+    private final WeakReference<Button> fetchAddressButtonReference;
+
+    public AddressResultReceiver(Handler handler, Context context, TextView resultTextView, ProgressBar progressBar,
+                                 Button fetchAddressButton) {
       super(handler);
+      this.contextReference = new WeakReference<>(context);
+      this.textViewReference = new WeakReference<>(resultTextView);
+      this.progressBarReference = new WeakReference<>(progressBar);
+      this.fetchAddressButtonReference = new WeakReference<>(fetchAddressButton);
     }
 
     /**
@@ -251,16 +282,28 @@ public class GeocodingServiceActivity extends AppCompatActivity {
     protected void onReceiveResult(int resultCode, Bundle resultData) {
       // Display the address string or an error message sent from the intent service.
       addressOutput = resultData.getString(Constants.RESULT_DATA_KEY);
-      displayAddressOutput();
+      TextView view = textViewReference.get();
+      if (view != null) {
+        displayAddressOutput(view);
+      }
 
       // Show a toast message if an address was found.
       if (resultCode == Constants.SUCCESS_RESULT) {
-        showToast(getString(R.string.address_found));
+        Context context = contextReference.get();
+        if (context != null) {
+          String message = context.getString(R.string.address_found);
+          showToast(context, message);
+        }
       }
 
       // Reset. Enable the Fetch Address button and stop showing the progress bar.
       addressRequested = false;
-      updateUiWidgets();
+
+      ProgressBar progressBar = progressBarReference.get();
+      Button fetchAddressButton = fetchAddressButtonReference.get();
+      if (progressBar != null && fetchAddressButton != null) {
+        updateUiWidgets(addressRequested, progressBar, fetchAddressButton);
+      }
     }
   }
 

--- a/mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingWidgetActivity.java
+++ b/mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingWidgetActivity.java
@@ -13,18 +13,17 @@ import com.mapbox.mapboxsdk.geometry.LatLng;
 import com.mapbox.mapboxsdk.maps.MapView;
 import com.mapbox.mapboxsdk.maps.MapboxMap;
 import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
+import com.mapbox.services.android.telemetry.location.LocationEngine;
+import com.mapbox.services.android.telemetry.location.LocationEngineListener;
 import com.mapbox.services.android.testapp.R;
 import com.mapbox.services.android.testapp.Utils;
+import com.mapbox.services.android.testapp.location.LostLocationEngine;
 import com.mapbox.services.android.ui.geocoder.GeocoderAutoCompleteView;
 import com.mapbox.services.api.geocoding.v5.GeocodingCriteria;
 import com.mapbox.services.api.geocoding.v5.models.CarmenFeature;
 import com.mapbox.services.commons.models.Position;
-import com.mapzen.android.lost.api.LocationListener;
-import com.mapzen.android.lost.api.LocationRequest;
-import com.mapzen.android.lost.api.LocationServices;
-import com.mapzen.android.lost.api.LostApiClient;
 
-public class GeocodingWidgetActivity extends AppCompatActivity {
+public class GeocodingWidgetActivity extends AppCompatActivity implements LocationEngineListener {
 
   private static final String LOG_TAG = "GeocodingWidgetActivity";
 
@@ -32,6 +31,7 @@ public class GeocodingWidgetActivity extends AppCompatActivity {
   private MapboxMap mapboxMap;
 
   private GeocoderAutoCompleteView autocomplete;
+  private LocationEngine locationEngine;
 
   @Override
   protected void onCreate(Bundle savedInstanceState) {
@@ -63,31 +63,24 @@ public class GeocodingWidgetActivity extends AppCompatActivity {
     });
 
     // Set up location services to improve accuracy
-    LocationRequest request = LocationRequest.create()
-      .setInterval(5000)
-      .setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
-    LostApiClient lostApiClient = new LostApiClient.Builder(this).build();
+    locationEngine = LostLocationEngine.getLocationEngine(this);
+    locationEngine.addLocationEngineListener(this);
+    locationEngine.activate();
+  }
 
-    //noinspection MissingPermission
-    LocationServices.FusedLocationApi.requestLocationUpdates(lostApiClient, request, new LocationListener() {
+  @Override
+  public void onConnected() {
+    Log.d(LOG_TAG, "Connected to engine, we can now request updates.");
+    locationEngine.requestLocationUpdates();
+  }
 
-      @Override
-      public void onLocationChanged(Location location) {
-        Log.d(LOG_TAG, "New LOST location: " + location.toString());
-        autocomplete.setProximity(Position.fromCoordinates(
-          location.getLongitude(), location.getLatitude()));
-      }
-
-      @Override
-      public void onProviderDisabled(String provider) {
-        Log.d(LOG_TAG, "onProviderDisabled: " + provider);
-      }
-
-      @Override
-      public void onProviderEnabled(String provider) {
-        Log.d(LOG_TAG, "onProviderEnabled: " + provider);
-      }
-    });
+  @Override
+  public void onLocationChanged(Location location) {
+    if (location != null) {
+      Log.d(LOG_TAG, "New LOST location: " + location.toString());
+      autocomplete.setProximity(Position.fromCoordinates(
+        location.getLongitude(), location.getLatitude()));
+    }
   }
 
   private void updateMap(double latitude, double longitude) {
@@ -114,12 +107,21 @@ public class GeocodingWidgetActivity extends AppCompatActivity {
   public void onResume() {
     super.onResume();
     mapView.onResume();
+    if (locationEngine != null && locationEngine.isConnected()) {
+      locationEngine.requestLocationUpdates();
+    }
   }
 
   @Override
   public void onPause() {
     super.onPause();
     mapView.onPause();
+    if (locationEngine != null) {
+      locationEngine.removeLocationUpdates();
+      locationEngine.removeLocationEngineListener(this);
+      locationEngine.deactivate();
+    }
+
   }
 
   @Override

--- a/mapbox/app/src/main/java/com/mapbox/services/android/testapp/location/GoogleLocationEngine.java
+++ b/mapbox/app/src/main/java/com/mapbox/services/android/testapp/location/GoogleLocationEngine.java
@@ -17,6 +17,8 @@ import com.mapbox.services.android.telemetry.location.LocationEngineListener;
 import com.mapbox.services.android.telemetry.location.LocationEnginePriority;
 import com.mapbox.services.android.telemetry.permissions.PermissionsManager;
 
+import java.lang.ref.WeakReference;
+
 /**
  * Sample LocationEngine using Google Play Services
  */
@@ -27,13 +29,13 @@ public class GoogleLocationEngine extends LocationEngine implements
 
   private static LocationEngine instance;
 
-  private Context context;
+  private WeakReference<Context> context;
   private GoogleApiClient googleApiClient;
 
   public GoogleLocationEngine(Context context) {
     super();
-    this.context = context;
-    googleApiClient = new GoogleApiClient.Builder(context)
+    this.context = new WeakReference<>(context);
+    googleApiClient = new GoogleApiClient.Builder(this.context.get())
       .addConnectionCallbacks(this)
       .addOnConnectionFailedListener(this)
       .addApi(LocationServices.API)
@@ -86,7 +88,7 @@ public class GoogleLocationEngine extends LocationEngine implements
 
   @Override
   public Location getLastLocation() {
-    if (googleApiClient.isConnected() && PermissionsManager.areLocationPermissionsGranted(context)) {
+    if (googleApiClient.isConnected() && PermissionsManager.areLocationPermissionsGranted(context.get())) {
       //noinspection MissingPermission
       return LocationServices.FusedLocationApi.getLastLocation(googleApiClient);
     }
@@ -112,7 +114,7 @@ public class GoogleLocationEngine extends LocationEngine implements
       request.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
     }
 
-    if (googleApiClient.isConnected() && PermissionsManager.areLocationPermissionsGranted(context)) {
+    if (googleApiClient.isConnected() && PermissionsManager.areLocationPermissionsGranted(context.get())) {
       //noinspection MissingPermission
       LocationServices.FusedLocationApi.requestLocationUpdates(googleApiClient, request, this);
     }

--- a/mapbox/app/src/main/java/com/mapbox/services/android/testapp/location/LocationEngineActivity.java
+++ b/mapbox/app/src/main/java/com/mapbox/services/android/testapp/location/LocationEngineActivity.java
@@ -44,6 +44,8 @@ public class LocationEngineActivity extends AppCompatActivity
     super.onPause();
     if (locationEngine != null) {
       locationEngine.removeLocationUpdates();
+      locationEngine.removeLocationEngineListener(this);
+      locationEngine.deactivate();
     }
   }
 

--- a/mapbox/app/src/main/java/com/mapbox/services/android/testapp/location/LostLocationEngine.java
+++ b/mapbox/app/src/main/java/com/mapbox/services/android/testapp/location/LostLocationEngine.java
@@ -13,6 +13,8 @@ import com.mapzen.android.lost.api.LocationRequest;
 import com.mapzen.android.lost.api.LocationServices;
 import com.mapzen.android.lost.api.LostApiClient;
 
+import java.lang.ref.WeakReference;
+
 /**
  * Sample LocationEngine using the Open Source Lost library
  */
@@ -23,13 +25,13 @@ public class LostLocationEngine extends LocationEngine implements
 
   private static LocationEngine instance;
 
-  private Context context;
+  private WeakReference<Context> context;
   private LostApiClient lostApiClient;
 
   public LostLocationEngine(Context context) {
     super();
-    this.context = context;
-    lostApiClient = new LostApiClient.Builder(context)
+    this.context = new WeakReference<>(context);
+    lostApiClient = new LostApiClient.Builder(this.context.get())
       .addConnectionCallbacks(this)
       .build();
   }
@@ -75,7 +77,7 @@ public class LostLocationEngine extends LocationEngine implements
 
   @Override
   public Location getLastLocation() {
-    if (lostApiClient.isConnected() && PermissionsManager.areLocationPermissionsGranted(context)) {
+    if (lostApiClient.isConnected() && PermissionsManager.areLocationPermissionsGranted(context.get())) {
       //noinspection MissingPermission
       return LocationServices.FusedLocationApi.getLastLocation(lostApiClient);
     }
@@ -101,7 +103,7 @@ public class LostLocationEngine extends LocationEngine implements
       request.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
     }
 
-    if (lostApiClient.isConnected() && PermissionsManager.areLocationPermissionsGranted(context)) {
+    if (lostApiClient.isConnected() && PermissionsManager.areLocationPermissionsGranted(context.get())) {
       //noinspection MissingPermission
       LocationServices.FusedLocationApi.requestLocationUpdates(lostApiClient, request, this);
     }

--- a/mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/location/AndroidLocationEngine.java
+++ b/mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/location/AndroidLocationEngine.java
@@ -9,6 +9,8 @@ import android.text.TextUtils;
 
 import com.mapbox.services.android.telemetry.permissions.PermissionsManager;
 
+import java.lang.ref.WeakReference;
+
 import timber.log.Timber;
 
 /**
@@ -23,7 +25,7 @@ public class AndroidLocationEngine extends LocationEngine implements LocationLis
 
   private static AndroidLocationEngine instance;
 
-  private Context context;
+  private WeakReference<Context> context;
   private LocationManager locationManager;
   private String currentProvider = null;
 
@@ -31,8 +33,8 @@ public class AndroidLocationEngine extends LocationEngine implements LocationLis
     super();
 
     Timber.v("Initializing.");
-    this.context = context;
-    locationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
+    this.context = new WeakReference<>(context);
+    locationManager = (LocationManager) this.context.get().getSystemService(Context.LOCATION_SERVICE);
     currentProvider = DEFAULT_PROVIDER;
 
   }
@@ -68,7 +70,7 @@ public class AndroidLocationEngine extends LocationEngine implements LocationLis
   @Override
   public Location getLastLocation() {
     if (!TextUtils.isEmpty(currentProvider)
-      && PermissionsManager.areLocationPermissionsGranted(context)) {
+      && PermissionsManager.areLocationPermissionsGranted(context.get())) {
       //noinspection MissingPermission
       return locationManager.getLastKnownLocation(currentProvider);
     }
@@ -79,7 +81,7 @@ public class AndroidLocationEngine extends LocationEngine implements LocationLis
   @Override
   public void requestLocationUpdates() {
     if (!TextUtils.isEmpty(currentProvider)
-      && PermissionsManager.areLocationPermissionsGranted(context)) {
+      && PermissionsManager.areLocationPermissionsGranted(context.get())) {
       //noinspection MissingPermission
       locationManager.requestLocationUpdates(currentProvider, DEFAULT_MIN_TIME, DEFAULT_MIN_DISTANCE, this);
     }
@@ -108,7 +110,7 @@ public class AndroidLocationEngine extends LocationEngine implements LocationLis
 
   @Override
   public void removeLocationUpdates() {
-    if (PermissionsManager.areLocationPermissionsGranted(context)) {
+    if (PermissionsManager.areLocationPermissionsGranted(context.get())) {
       //noinspection MissingPermission
       locationManager.removeUpdates(this);
     }

